File: web/src/lib/utils/formatting.ts
import Decimal from 'decimal.js';

export function formatCurrency(amount: string | number | Decimal, symbol = '£'): string {
  const value = new Decimal(amount.toString());
  const abs = value.abs().toFixed(2);
  const formatted = abs.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  return value.isNeg() ? `-${symbol}${formatted}` : `${symbol}${formatted}`;
}

export function formatPercentage(value: number, decimals = 1): string {
  return `${value.toFixed(decimals)}%`;
}

File: finadviser/utils/formatting.py
"""Formatting utilities for display."""

from __future__ import annotations

from decimal import Decimal


def format_currency(amount: Decimal | float, symbol: str = "$") -> str:
    """Format a numeric amount as currency."""
    value = Decimal(str(amount))
    if value < 0:
        return f"-{symbol}{abs(value):,.2f}"
    return f"{symbol}{value:,.2f}"


def format_percentage(value: float, decimals: int = 1) -> str:
    """Format a float as a percentage string."""
    return f"{value:.{decimals}f}%"
